packages/diagram/src/layout/engine/layout.ts

```diff
@@ -0,0 +1,426 @@
+import { StyleList, Selector, SelectorType, Style } from "../../styles.js";
+import {
+    LayoutElement,
+    LayoutInformation,
+    SizeConstraints,
+    addToSize,
+    matchToConstraints,
+    HorizontalAlignment,
+    VerticalAlignment
+} from "../layoutElement.js";
+import { LayoutEngine } from "./layoutEngine.js";
+import { Element } from "@hylimo/diagram-common";
+import { generateEdits } from "./edits.js";
+import { CanvasLayoutEngine } from "./canvasLayoutEngine.js";
+
+/**
+ * Performs the layout, uses a layout engine to do so
+ */
+
+export class Layout {
+    /**
+     * Lookup for layout elements
+     */
+    readonly layoutElementLookup = new Map<string, LayoutElement>();
+
+    /**
+     * Lookup for layouted elements
+     */
+    readonly elementLookup: Record<string, Element> = {};
+
+    /**
+     * Lookop for element id based on syncscript object
+     */
+    readonly elementIdLookup: Map<FullObject, string> = new Map();
+
+    /**
+     * Counter to provide child ids
+     */
+    private readonly elementIdCounter: Map<string, number> = new Map();
+
+    /**
+     * Canvas layout engine to layout all canvases
+     */
+    readonly layoutEngine = new CanvasLayoutEngine(this);
+
+    /**
+     * Creates a new layout
+     *
+     * @param engine the engine which provides fonts
+     * @param styles styles to possibly apply to elements
+     * @param fonts fonts to use
+     * @param defaultFont the default font to use
+     */
+    constructor(
+        readonly engine: LayoutEngine,
+        readonly styles: StyleList,
+        readonly fonts: FontCollection,
+        readonly defaultFont: FontFamily
+    ) {}
+
+    /**
+     * Checks if an element matches a selector
+     * @param element the element to check
+     * @param selector the selector to check
+     * @returns true if the element matches the selector, otherwise false
+     */
+    private matchesSelector(element: LayoutElement, selector: Selector): boolean {
+        if (selector.type === SelectorType.CLASS) {
+            return element.class.has(selector.value);
+        } else if (selector.type === SelectorType.TYPE) {
+            return element.layoutConfig.type === selector.value;
+        } else if (selector.type === SelectorType.ANY) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Checks if a style should be applied to an element
+     *
+     * @param element the element to which the style may be applied
+     * @param style the style to check
+     * @returns true if the style should be applied
+     */
+    private matchesStyle(element: LayoutElement, style: Style): boolean {
+        let currentElement: LayoutElement | undefined = element;
+        const initialElementIndex = style.selectorChain.length - 1;
+        let i = initialElementIndex;
+
+        while (currentElement) {
+            if (this.matchesSelector(currentElement, style.selectorChain[i])) {
```

Comment:
Ouch.
It is unfortunate that you need to reimplement CSS in Hylimo, but I can imagine why it is necessary…

Reply:
when I originally did this, I also searched for some lib that already does this and which can be used in a somewhat performant way
back then I didn't find anything I liked
but I guess at this point it doesn't really matter anymore, even if we find sth as the semantics are not 100% identical